### PR TITLE
Raspberrypi cluster

### PR DESCRIPTION
**Pull request by twerthi**

### CODE REVIEW

File: blog/2020-05/raspberrypi-cluster/index.md [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+published: 2999-01-01
+metaImage:
+bannerImage:
+tags:
+ -
+---
+
+I recently decided to use Docker to replace the VMs on my hypervisor with containers. In this post, I share some of the lessons I learned running the new Raspberry Pi 4 in a cluster with Docker Swarm.
+
+## Parts
```
Comment (robpearson) [inline blog/2020-05/raspberrypi-cluster/index.md:15]: I'd make this more explicit. `Raspberry PI hardware`

File: blog/2020-05/raspberrypi-cluster/index.md [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+author: shawn.sesna@octopus.com
+visibility: private
+published: 2999-01-01
+metaImage:
+bannerImage:
+tags:
+ -
+---
+
+I recently decided to use Docker to replace the VMs on my hypervisor with containers. In this post, I share some of the lessons I learned running the new Raspberry Pi 4 in a cluster with Docker Swarm.
```
Comment (robpearson) [inline blog/2020-05/raspberrypi-cluster/index.md:13]: I drafted a suggestion to try to blend your old one and a BLUF intro line plus a bit of WHY.

---

Like many others, I had to stay home through the COVID-19 crisis, so I decided to build a Raspberry PI cluster to explore running applications on it with Docker Swarm. In this post, I share some of the lessons I learned running the new Raspberry Pi 4 in a cluster with Docker Swarm.

## Why 

I'm a developer, and I've been interested in starting a Raspberry PI project for a while. I'm also interested in cluster computing, but never had a good reason to create a cluster. Then I read about people having great success running the new Raspberry Pi 4 in a cluster with Docker Swarm, so I thought it was an opportune time to learn more.

I kicked off this project with several goals:
* Have fun as Raspberry PIs are pretty cool hardware.
* Shift some computing resources from a home hypervisor server running VMs to the small, power-efficient Raspberry PI cluster running containers.
* Learn Docker Swarm.

File: blog/2020-05/raspberrypi-cluster/index.md [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+Lastly, I needed a network switch to connect the Pi devices to, but I had a spare 8-port, so I didn’t need to buy another.  The Raspberry Pi 4 comes with a wireless adapter built-in, so you could utilize that as well.
+
+## Creating the cluster
+
+There are tons of blog posts out there that show you how to create a cluster of Raspberry Pi with Docker Swarm.  I found [How to run a Raspberry Pi cluster with Docker Swarm](https://howchoo.com/g/njy4zdm3mwy/how-to-run-a-raspberry-pi-cluster-with-docker-swarm) was clear and easy to follow.  All of the details you need are in that post.
+
+## Operating the Swarm
+
+I have played around with Kubernetes (K8s) and Docker before taking on this project. However, I’d never run a Docker Swarm.  There were some new concepts to learn when running things in Docker Swarm mode versus a stand-alone Docker instance.  For example, where you would normally use a `docker run` statement, you’d create a service within the swarm to run the container using `docker service create`.  After messing around for a while, I had a fully operational Swarm:
```
Comment (robpearson) [inline blog/2020-05/raspberrypi-cluster/index.md:43]: I'd probably say something other than played around ... :)  I `have experience with` Kubernetes (K8s) and Docker ...

File: blog/2020-05/raspberrypi-cluster/index.md [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+In some cases, you can use a work-around and tell Docker not to resolve the image to an architecture:
+
+- `docker service create --name <somename> --no-resolve-image <image>`
+- `docker stack deploy --name <somename> --compose-file <somefile> --resolve-image never`
+
+## Lesson 2: Stacks
+
+Stacks are the Docker Swarm way of using docker-compose. In fact, it uses a compose file as one of its arguments.  Just like docker-compose, you can define multiple containers within a single file.  By default, a stack deployment will create a network for the stack so they can all talk to each other.  This makes it easy to refer one container to another just by name.  For example, below is a compose file for running [Home Assistant](https://www.home-assistant.io/) on my docker swarm.  This stack consists of an MQTT service container, a Home Assistant container, and a MySQL database server container.  For reasons I’ll go into later, I configured Home Assistant to use a MySQL back-end for the Recorder:
```
Comment (robpearson) [inline blog/2020-05/raspberrypi-cluster/index.md:58]: Probably good to add a few links for help on some of these topics. Just for more context. Docker Compose. Any helpful links.

File: blog/2020-05/raspberrypi-cluster/index.md [hunk 5]
```diff
@@ -100,7 +109,7 @@ networks:
   hass:
 ```
 
-In the `configuration.yaml` file, you can see where I created the connection to the mysqldb container:
+In the `configuration.yaml` of file of Home Assistant, you can see where I created the connection to the mysqldb container:
```
Comment (wordlee) [inline blog/2020-05/raspberrypi-cluster/index.md:112]:
```suggestion
In the `configuration.yaml` file of Home Assistant, you can see where I created the connection to the mysqldb container:
```

**Pull request merged**